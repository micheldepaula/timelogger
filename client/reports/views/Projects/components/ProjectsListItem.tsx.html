
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for views/Projects/components/ProjectsListItem.tsx</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../../index.html">All files</a> / <a href="index.html">views/Projects/components</a> ProjectsListItem.tsx</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">50% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>10/20</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">50% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>5/10</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">11.11% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>1/9</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">50% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>10/20</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input oninput="onInput()" type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line medium'></div>
    <pre><table class="coverage">
<tr><td class="line-count quiet"><a name='L1'></a><a href='#L1'>1</a>
<a name='L2'></a><a href='#L2'>2</a>
<a name='L3'></a><a href='#L3'>3</a>
<a name='L4'></a><a href='#L4'>4</a>
<a name='L5'></a><a href='#L5'>5</a>
<a name='L6'></a><a href='#L6'>6</a>
<a name='L7'></a><a href='#L7'>7</a>
<a name='L8'></a><a href='#L8'>8</a>
<a name='L9'></a><a href='#L9'>9</a>
<a name='L10'></a><a href='#L10'>10</a>
<a name='L11'></a><a href='#L11'>11</a>
<a name='L12'></a><a href='#L12'>12</a>
<a name='L13'></a><a href='#L13'>13</a>
<a name='L14'></a><a href='#L14'>14</a>
<a name='L15'></a><a href='#L15'>15</a>
<a name='L16'></a><a href='#L16'>16</a>
<a name='L17'></a><a href='#L17'>17</a>
<a name='L18'></a><a href='#L18'>18</a>
<a name='L19'></a><a href='#L19'>19</a>
<a name='L20'></a><a href='#L20'>20</a>
<a name='L21'></a><a href='#L21'>21</a>
<a name='L22'></a><a href='#L22'>22</a>
<a name='L23'></a><a href='#L23'>23</a>
<a name='L24'></a><a href='#L24'>24</a>
<a name='L25'></a><a href='#L25'>25</a>
<a name='L26'></a><a href='#L26'>26</a>
<a name='L27'></a><a href='#L27'>27</a>
<a name='L28'></a><a href='#L28'>28</a>
<a name='L29'></a><a href='#L29'>29</a>
<a name='L30'></a><a href='#L30'>30</a>
<a name='L31'></a><a href='#L31'>31</a>
<a name='L32'></a><a href='#L32'>32</a>
<a name='L33'></a><a href='#L33'>33</a>
<a name='L34'></a><a href='#L34'>34</a>
<a name='L35'></a><a href='#L35'>35</a>
<a name='L36'></a><a href='#L36'>36</a>
<a name='L37'></a><a href='#L37'>37</a>
<a name='L38'></a><a href='#L38'>38</a>
<a name='L39'></a><a href='#L39'>39</a>
<a name='L40'></a><a href='#L40'>40</a>
<a name='L41'></a><a href='#L41'>41</a>
<a name='L42'></a><a href='#L42'>42</a>
<a name='L43'></a><a href='#L43'>43</a>
<a name='L44'></a><a href='#L44'>44</a>
<a name='L45'></a><a href='#L45'>45</a>
<a name='L46'></a><a href='#L46'>46</a>
<a name='L47'></a><a href='#L47'>47</a>
<a name='L48'></a><a href='#L48'>48</a>
<a name='L49'></a><a href='#L49'>49</a>
<a name='L50'></a><a href='#L50'>50</a>
<a name='L51'></a><a href='#L51'>51</a>
<a name='L52'></a><a href='#L52'>52</a>
<a name='L53'></a><a href='#L53'>53</a>
<a name='L54'></a><a href='#L54'>54</a>
<a name='L55'></a><a href='#L55'>55</a>
<a name='L56'></a><a href='#L56'>56</a>
<a name='L57'></a><a href='#L57'>57</a>
<a name='L58'></a><a href='#L58'>58</a>
<a name='L59'></a><a href='#L59'>59</a>
<a name='L60'></a><a href='#L60'>60</a>
<a name='L61'></a><a href='#L61'>61</a>
<a name='L62'></a><a href='#L62'>62</a>
<a name='L63'></a><a href='#L63'>63</a>
<a name='L64'></a><a href='#L64'>64</a>
<a name='L65'></a><a href='#L65'>65</a>
<a name='L66'></a><a href='#L66'>66</a>
<a name='L67'></a><a href='#L67'>67</a>
<a name='L68'></a><a href='#L68'>68</a>
<a name='L69'></a><a href='#L69'>69</a>
<a name='L70'></a><a href='#L70'>70</a>
<a name='L71'></a><a href='#L71'>71</a>
<a name='L72'></a><a href='#L72'>72</a>
<a name='L73'></a><a href='#L73'>73</a>
<a name='L74'></a><a href='#L74'>74</a>
<a name='L75'></a><a href='#L75'>75</a>
<a name='L76'></a><a href='#L76'>76</a>
<a name='L77'></a><a href='#L77'>77</a>
<a name='L78'></a><a href='#L78'>78</a>
<a name='L79'></a><a href='#L79'>79</a>
<a name='L80'></a><a href='#L80'>80</a>
<a name='L81'></a><a href='#L81'>81</a>
<a name='L82'></a><a href='#L82'>82</a>
<a name='L83'></a><a href='#L83'>83</a>
<a name='L84'></a><a href='#L84'>84</a>
<a name='L85'></a><a href='#L85'>85</a>
<a name='L86'></a><a href='#L86'>86</a>
<a name='L87'></a><a href='#L87'>87</a>
<a name='L88'></a><a href='#L88'>88</a>
<a name='L89'></a><a href='#L89'>89</a>
<a name='L90'></a><a href='#L90'>90</a>
<a name='L91'></a><a href='#L91'>91</a>
<a name='L92'></a><a href='#L92'>92</a>
<a name='L93'></a><a href='#L93'>93</a>
<a name='L94'></a><a href='#L94'>94</a>
<a name='L95'></a><a href='#L95'>95</a>
<a name='L96'></a><a href='#L96'>96</a>
<a name='L97'></a><a href='#L97'>97</a>
<a name='L98'></a><a href='#L98'>98</a>
<a name='L99'></a><a href='#L99'>99</a>
<a name='L100'></a><a href='#L100'>100</a>
<a name='L101'></a><a href='#L101'>101</a></td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">21x</span>
<span class="cline-any cline-yes">21x</span>
<span class="cline-any cline-yes">21x</span>
<span class="cline-any cline-yes">21x</span>
<span class="cline-any cline-yes">21x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">21x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">21x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">21x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">21x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import { useAppDispatch } from '../../../hooks';
import React, { useState } from 'react';
import { IProject } from '../../../../utils';
import { editProject } from '../../../slice/projectSlice';
&nbsp;
type ProjectsListItemProps = {
  project: IProject;
};
&nbsp;
const ProjectsListItem: React.FC&lt;ProjectsListItemProps&gt; = ({ project }) =&gt; {
  const [edit, setEdit] = useState&lt;boolean&gt;(false);
  const [hours, setHours] = useState(project.hours);
  const [name, setName] = useState(project.name);
  const dispatch = useAppDispatch();
  const handleEditProject = <span class="fstat-no" title="function not covered" >()</span> =&gt; {
<span class="cstat-no" title="statement not covered" >    setEdit(false);</span>
<span class="cstat-no" title="statement not covered" >    if (project.end) {</span>
<span class="cstat-no" title="statement not covered" >      dispatch(</span>
        editProject({
          id: project.id,
          name: name,
          hours: Number(hours),
          end: new Date(project.end.toString()),
        }),
      );
    }
  };
  const EditableInputName = <span class="fstat-no" title="function not covered" >()</span> =&gt; {
<span class="cstat-no" title="statement not covered" >    return (</span>
      &lt;td className='border px-4 py-2'&gt;
        &lt;input
          className='border rounded-full py-2 px-4'
          value={name}
          type='name'
          placeholder='Project Name'
          aria-label='Project Name'
          onChange={<span class="fstat-no" title="function not covered" >(e</span>) =&gt; <span class="cstat-no" title="statement not covered" >setName(e.target.value.toString())}</span>
        /&gt;
      &lt;/td&gt;
    );
  };
&nbsp;
  const EditableInputHours = <span class="fstat-no" title="function not covered" >()</span> =&gt; {
<span class="cstat-no" title="statement not covered" >    return (</span>
      &lt;td className='border px-4 py-2 '&gt;
        &lt;input
          className='border rounded-full py-2 px-4'
          value={hours}
          type='number'
          placeholder='DeadLine'
          aria-label='DeadLine'
          onChange={<span class="fstat-no" title="function not covered" >(e</span>) =&gt; <span class="cstat-no" title="statement not covered" >setHours(e.target.value.toString())}</span>
        /&gt;
      &lt;/td&gt;
    );
  };
&nbsp;
  const EditButtons = <span class="fstat-no" title="function not covered" >()</span> =&gt; {
<span class="cstat-no" title="statement not covered" >    return (</span>
      &lt;div className='flex justify-between '&gt;
        &lt;button
          className='bg-blue-500 hover:bg-blue-700 text-white rounded-full py-2 px-4 ml-2'
          onClick={handleEditProject}
          data-testid={`projects-list-item-${project.id}-button-edit`}
        &gt;
          Save
        &lt;/button&gt;
        &lt;button
          className='bg-red-500 hover:bg-red-700 text-white rounded-full py-2 px-4 ml-2'
          onClick={<span class="fstat-no" title="function not covered" >()</span> =&gt; <span class="cstat-no" title="statement not covered" >setEdit(false)}</span>
          data-testid={`projects-list-item-${project.id}-button-cancel-edit`}
        &gt;
          cancel
        &lt;/button&gt;
      &lt;/div&gt;
    );
  };
&nbsp;
  return (
    &lt;tr data-testid={`projects-list-item-${project.id}`}&gt;
      &lt;td className='border px-4 py-2 w-12'&gt;{project.id}&lt;/td&gt;
      {!edit ? &lt;td className='border px-4 py-2'&gt;{name}&lt;/td&gt; : <span class="branch-1 cbranch-no" title="branch not covered" >&lt;EditableInputName /&gt;}</span>
      {!edit ? &lt;td className='border px-4 py-2'&gt;{hours}&lt;/td&gt; : <span class="branch-1 cbranch-no" title="branch not covered" >&lt;EditableInputHours /&gt;}</span>
      &lt;td className='flex justify-between border px-4 py-2 '&gt;
        {project.end &amp;&amp; new Date(project.end).toISOString().split('T')[0]}
        {!edit ? (
          &lt;button
            className='bg-blue-500 hover:bg-blue-700 text-white rounded-full py-2 px-4 ml-2'
            onClick={<span class="fstat-no" title="function not covered" >()</span> =&gt; <span class="cstat-no" title="statement not covered" >setEdit(true)}</span>
          &gt;
            Edit
          &lt;/button&gt;
        ) : (
<span class="branch-1 cbranch-no" title="branch not covered" >          &lt;EditButtons /&gt;</span>
        )}
      &lt;/td&gt;
    &lt;/tr&gt;
  );
};
export default ProjectsListItem;
&nbsp;</pre></td></tr></table></pre>

                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2023-06-30T10:46:42.873Z
            </div>
        <script src="../../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../../sorter.js"></script>
        <script src="../../../block-navigation.js"></script>
    </body>
</html>
    