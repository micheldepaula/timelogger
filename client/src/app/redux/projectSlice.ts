import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit'
import { RootState } from '../../app/store'
import { getAll, postProject, putProject } from './../api/projects'

export interface Project {
  id?: number
  name?: string
  hours?: number
  end: Date
}

export interface ProjectState {
  loading: boolean
  projects: Array<Project>
  project: Project | undefined
  error: string | undefined
}
const initialState: ProjectState = {
  loading: false,
  projects: [],
  project: undefined,
  error: undefined,
}
export const getProjects = createAsyncThunk('projects/getProjects', () => {
  const res = getAll().then((data) => data)
  return res
})

export const addProject = createAsyncThunk('projects/addProjects', (project: Project) => {
  const res = postProject(project).then((data) => data)
  return res
})

export const editProject = createAsyncThunk('projects/editProjects', (project: Project) => {
  const res = putProject(project).then((data) => data)
  return res
})

export const sortedProjectsAsc = createAsyncThunk('projects/sortedProjectsAsc', (projects: Array<Project>) => {
  const res = [...projects].sort((a, b) => new Date(b.end).getTime() - new Date(a.end).getTime())
  return res
})

export const sortedProjectsDesc2 = createAsyncThunk('projects/sortedProjectsDesc2', (projects: Array<Project>) => {
  const res = [...projects].sort((a, b) => new Date(a.end).getTime() - new Date(b.end).getTime())
  return res
})

const ProjectSlice = createSlice({
  name: 'projects',
  initialState,
  extraReducers: (builder) => {
    builder.addCase(getProjects.pending, (state) => {
      state.loading = true
    })
    builder.addCase(getProjects.fulfilled, (state, action: PayloadAction<Array<Project>>) => {
      state.loading = false
      state.projects = action.payload
    })
    builder.addCase(getProjects.rejected, (state, action) => {
      state.loading = false
      state.projects = []
      state.error = action.error.message
    })
    builder.addCase(addProject.pending, (state) => {
      state.loading = true
    })
    builder.addCase(addProject.fulfilled, (state, action: PayloadAction<Project>) => {
      state.loading = false
      state.projects = [...state.projects, action.payload]
    })
    builder.addCase(addProject.rejected, (state, action) => {
      state.loading = false
      state.error = action.error.message
    })
    builder.addCase(editProject.pending, (state) => {
      state.loading = true
    })
    builder.addCase(editProject.fulfilled, (state, action: PayloadAction<Project>) => {
      state.loading = false
      state.project = action.payload
    })
    builder.addCase(editProject.rejected, (state, action) => {
      state.loading = false
      state.error = action.error.message
    })
    builder.addCase(sortedProjectsAsc.pending, (state) => {
      state.loading = false
    })

    builder.addCase(sortedProjectsAsc.fulfilled, (state, action: PayloadAction<Array<Project>>) => {
      state.loading = false
      state.projects = action.payload
    })

    builder.addCase(sortedProjectsAsc.rejected, (state, action) => {
      state.loading = false
      state.error = action.error.message
    })
    builder.addCase(sortedProjectsDesc2.pending, (state) => {
      state.loading = false
    })

    builder.addCase(sortedProjectsDesc2.fulfilled, (state, action: PayloadAction<Array<Project>>) => {
      state.loading = false
      state.projects = action.payload
    })

    builder.addCase(sortedProjectsDesc2.rejected, (state, action) => {
      state.loading = false
      state.error = action.error.message
    })
  },
  reducers: {
    sortedProjectsDesc: (state) => {
      state.projects = state.projects.sort((a, b) => new Date(a.end).getTime() - new Date(b.end).getTime())
    },
  },
})

export const { sortedProjectsDesc } = ProjectSlice.actions
export const projectsSelector = (state: RootState) => state.projectsReducer
export default ProjectSlice.reducer
