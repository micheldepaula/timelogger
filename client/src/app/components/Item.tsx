import { useAppDispatch } from '../../app/hooks'
import React, { useState } from 'react'
import { IProject } from '../../utils'
import { editProject } from '../redux/projectSlice'

type ItemProps = {
  project: IProject
}

export default function Item(props: ItemProps) {
  const [edit, setEdit] = useState<boolean>(false)
  const [hours, setHours] = useState(props.project.hours)
  const [name, setName] = useState(props.project.name)
  const dispatch = useAppDispatch()
  const handleEditProject = () => {
    setEdit(false)
    if (props.project.end) {
      dispatch(
        editProject({
          id: props.project.id,
          name: name,
          hours: Number(hours),
          end: new Date(props.project.end.toString()),
        }),
      )
    }
  }
  const EditableInputName = () => {
    return (
      <td className='border px-4 py-2'>
        <input
          className='border rounded-full py-2 px-4'
          value={name}
          type='name'
          placeholder='Project Name'
          aria-label='Project Name'
          onChange={(e) => setName(e.target.value.toString())}
        />
      </td>
    )
  }

  const EditableInputHours = () => {
    return (
      <td className='border px-4 py-2 '>
        <input
          className='border rounded-full py-2 px-4'
          value={hours}
          type='number'
          placeholder='DeadLine'
          aria-label='DeadLine'
          onChange={(e) => setHours(e.target.value.toString())}
        />
      </td>
    )
  }

  const EditButtons = () => {
    return (
      <div className='flex justify-between '>
        <button
          className='bg-blue-500 hover:bg-blue-700 text-white rounded-full py-2 px-4 ml-2'
          onClick={handleEditProject}
        >
          Save
        </button>
        <button
          className='bg-red-500 hover:bg-red-700 text-white rounded-full py-2 px-4 ml-2'
          onClick={() => setEdit(false)}
        >
          cancel
        </button>
      </div>
    )
  }

  return (
    <tr>
      <td className='border px-4 py-2 w-12'>{props.project.id}</td>
      {!edit ? <td className='border px-4 py-2'>{name}</td> : <EditableInputName />}
      {!edit ? <td className='border px-4 py-2'>{hours}</td> : <EditableInputHours />}
      <td className='flex justify-between border px-4 py-2 '>
        {props.project.end && new Date(props.project.end).toISOString().split('T')[0]}
        {!edit ? (
          <button
            className='bg-blue-500 hover:bg-blue-700 text-white rounded-full py-2 px-4 ml-2'
            onClick={() => setEdit(true)}
          >
            Edit
          </button>
        ) : (
          <EditButtons />
        )}
      </td>
    </tr>
  )
}
