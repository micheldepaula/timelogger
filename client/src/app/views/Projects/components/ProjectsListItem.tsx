import { useAppDispatch } from '../../../hooks'
import React, { useState } from 'react'
import { IProject } from '../../../../utils'
import { editProject } from '../../../slice/projectSlice'

type ProjectsListItemProps = {
  project: IProject
}

const ProjectsListItem: React.FC<ProjectsListItemProps> = ({ project }) => {
  const [edit, setEdit] = useState<boolean>(false)
  const [hours, setHours] = useState(project.hours)
  const [name, setName] = useState(project.name)
  const dispatch = useAppDispatch()
  const handleEditProject = () => {
    setEdit(false)
    if (project.end) {
      dispatch(
        editProject({
          id: project.id,
          name: name,
          hours: Number(hours),
          end: new Date(project.end.toString()),
        }),
      )
    }
  }
  const EditableInputName = () => {
    return (
      <td className='border px-4 py-2'>
        <input
          className='border rounded-full py-2 px-4'
          value={name}
          type='name'
          placeholder='Project Name'
          aria-label='Project Name'
          onChange={(e) => setName(e.target.value.toString())}
        />
      </td>
    )
  }

  const EditableInputHours = () => {
    return (
      <td className='border px-4 py-2 '>
        <input
          className='border rounded-full py-2 px-4'
          value={hours}
          type='number'
          placeholder='DeadLine'
          aria-label='DeadLine'
          onChange={(e) => setHours(e.target.value.toString())}
        />
      </td>
    )
  }

  const EditButtons = () => {
    return (
      <div className='flex justify-between '>
        <button
          className='bg-blue-500 hover:bg-blue-700 text-white rounded-full py-2 px-4 ml-2'
          onClick={handleEditProject}
          data-testid={`projects-list-item-${project.id}-button-edit`}
        >
          Save
        </button>
        <button
          className='bg-red-500 hover:bg-red-700 text-white rounded-full py-2 px-4 ml-2'
          onClick={() => setEdit(false)}
          data-testid={`projects-list-item-${project.id}-button-cancel-edit`}
        >
          cancel
        </button>
      </div>
    )
  }

  return (
    <tr data-testid={`projects-list-item-${project.id}`}>
      <td className='border px-4 py-2 w-12'>{project.id}</td>
      {!edit ? <td className='border px-4 py-2'>{name}</td> : <EditableInputName />}
      {!edit ? <td className='border px-4 py-2'>{hours}</td> : <EditableInputHours />}
      <td className='flex justify-between border px-4 py-2 '>
        {project.end && new Date(project.end).toISOString().split('T')[0]}
        {!edit ? (
          <button
            className='bg-blue-500 hover:bg-blue-700 text-white rounded-full py-2 px-4 ml-2'
            onClick={() => setEdit(true)}
          >
            Edit
          </button>
        ) : (
          <EditButtons />
        )}
      </td>
    </tr>
  )
}
export default ProjectsListItem
