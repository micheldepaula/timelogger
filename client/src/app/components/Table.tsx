import React, { useEffect, useState } from 'react'
import Item from './Item'

import { useAppDispatch, useAppSelector } from '../../app/hooks'
import { sortedProjectsAsc, projectsSelector, getProjects } from './../redux/projectSlice'

export default function Table() {
  const [up, setUp] = useState<boolean>(true)
  const dispatch = useAppDispatch()

  const globalData = useAppSelector(projectsSelector)

  useEffect(() => {
    dispatch(getProjects())
  }, [])

  const orderProjectsByEndDate = () => {
    if (up) {
      dispatch(sortedProjectsAsc(globalData.projects))
      setUp(!up)
    } else {
      dispatch({ type: 'projects/sortedProjectsDesc' })
      setUp(!up)
    }
  }

  const IconUp = () => {
    return (
      <svg
        xmlns='http://www.w3.org/2000/svg'
        fill='none'
        viewBox='0 0 24 24'
        strokeWidth={1.5}
        stroke='currentColor'
        className='w-6 h-6'
      >
        <path strokeLinecap='round' strokeLinejoin='round' d='M12 19.5v-15m0 0l-6.75 6.75M12 4.5l6.75 6.75' />
      </svg>
    )
  }
  const IconDown = () => {
    return (
      <svg
        xmlns='http://www.w3.org/2000/svg'
        fill='none'
        viewBox='0 0 24 24'
        strokeWidth={1.5}
        stroke='currentColor'
        className='w-6 h-6'
      >
        <path strokeLinecap='round' strokeLinejoin='round' d='M19.5 13.5L12 21m0 0l-7.5-7.5M12 21V3' />
      </svg>
    )
  }

  return (
    <table className='table-fixed w-full'>
      <thead className='bg-gray-200'>
        <tr className='space'>
          <th className='border px-4 py-2 w-12'>#</th>
          <th className='border px-4 py-2'>Project Name</th>
          <th className='border px-4 py-2'>Hours</th>
          <th className='flex justify-between border px-4 py-2 cursor-pointer ' onClick={orderProjectsByEndDate}>
            DeadLine {up ? <IconUp /> : <IconDown />}
          </th>
        </tr>
      </thead>
      <tbody>
        {globalData.loading && <div>Loading...</div>}
        {globalData.error && <div>Error: {globalData.error}</div>}
        {globalData.projects.map((project) => (
          <Item key={project.id} project={project} />
        ))}
      </tbody>
    </table>
  )
}
