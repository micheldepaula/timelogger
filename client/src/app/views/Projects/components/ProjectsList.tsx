import React, { useState } from 'react'
import Item from './ProjectsListItem'

import { useAppDispatch } from '../../../hooks'
import { sortedProjectsAsc, Project } from '../../../slice/projectSlice'

interface ProjectsListProps {
  projects: Array<Project>
  error?: string
  loading?: string
}
const ProjectsList: React.FC<ProjectsListProps> = ({ projects, error, loading }) => {
  const [up, setUp] = useState<boolean>(true)
  const dispatch = useAppDispatch()

  const orderProjectsByEndDate = () => {
    if (up) {
      dispatch(sortedProjectsAsc(projects))
      setUp(!up)
    } else {
      dispatch({ type: 'projects/sortedProjectsDesc' })
      setUp(!up)
    }
  }

  const IconUp = () => {
    return (
      <svg
        xmlns='http://www.w3.org/2000/svg'
        fill='none'
        viewBox='0 0 24 24'
        strokeWidth={1.5}
        stroke='currentColor'
        className='w-6 h-6'
      >
        <path strokeLinecap='round' strokeLinejoin='round' d='M12 19.5v-15m0 0l-6.75 6.75M12 4.5l6.75 6.75' />
      </svg>
    )
  }
  const IconDown = () => {
    return (
      <svg
        xmlns='http://www.w3.org/2000/svg'
        fill='none'
        viewBox='0 0 24 24'
        strokeWidth={1.5}
        stroke='currentColor'
        className='w-6 h-6'
      >
        <path strokeLinecap='round' strokeLinejoin='round' d='M19.5 13.5L12 21m0 0l-7.5-7.5M12 21V3' />
      </svg>
    )
  }

  return (
    <table className='table-fixed w-full' data-testid='projects-list-component'>
      <thead className='bg-gray-200'>
        <tr className='space'>
          <th className='border px-4 py-2 w-12'>#</th>
          <th className='border px-4 py-2'>Project Name</th>
          <th className='border px-4 py-2'>Hours</th>
          <th className='flex justify-between border px-4 py-2 cursor-pointer ' onClick={orderProjectsByEndDate}>
            DeadLine {up ? <IconUp /> : <IconDown />}
          </th>
        </tr>
      </thead>
      <tbody>
        {loading && <div>Loading...</div>}
        {error && <div>Error: {error}</div>}
        {projects.map((project) => (
          <Item key={project.id} project={project} />
        ))}
      </tbody>
    </table>
  )
}

export default ProjectsList
