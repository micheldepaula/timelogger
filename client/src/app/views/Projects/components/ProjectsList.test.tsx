import React from 'react';
import { render } from '@testing-library/react';
import { Wrapper } from './../../../components';
import ProjectsList from './ProjectsList';
import { Project } from '../../../slice/projectSlice';
jest.mock('react-redux', () => {
  const actualModule = jest.requireActual('react-redux');
  return {
    __esModule: true,
    ...actualModule,
  };
});
describe('Projects List Component', () => {
  const projects: Array<Project> = [];
  for (let i = 1; i <= 10; i++) {
    projects.push({ end: new Date('2023-06-29'), hours: 8, id: i, name: `Project List Item ${i}` });
  }

  const rerender = () =>
    render(
      <Wrapper>
        <ProjectsList projects={projects} />
      </Wrapper>,
    );

  it('renders correctly', () => {
    const { asFragment, getByTestId } = rerender();
    expect(getByTestId('projects-list-component')).toBeDefined();
    projects.map((project) => expect(getByTestId(`projects-list-item-${project.id}`)).toBeDefined());
    expect(asFragment).toMatchSnapshot();
  });

  it('should order the projects end date', () => {
    const { asFragment, getByTestId } = rerender();
    expect(getByTestId('projects-list-component')).toBeDefined();
    projects.map((project) => expect(getByTestId(`projects-list-item-${project.id}`)).toBeDefined());

    expect(asFragment).toMatchSnapshot();
  });
});
