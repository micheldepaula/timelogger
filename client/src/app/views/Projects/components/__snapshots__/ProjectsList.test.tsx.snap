// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projects List Component renders correctly 1`] = `[Function]`;
